packages/@ember/-internals/glimmer/lib/syntax/outlet.ts

```diff
@@ -85,17 +92,94 @@ export function outletMacro(
   return true;
 }
 
+class OutletModelReference implements VersionedPathReference {
+  public tag: Tag;
+
+  constructor(private parent: VersionedPathReference<OutletState | undefined>) {
+    this.tag = parent.tag;
+  }
+
+  value(): Opaque {
+    let state = this.parent.value();
+
+    if (state === undefined) {
+      return undefined;
+    }
+
+    let { render } = state;
+
+    if (render === undefined) {
+      return undefined;
+    }
+
+    return render.model as Opaque;
+  }
+
+  get(property: string): VersionedPathReference {
+    if (DEBUG) {
+      // This guarentees that we preserve the `debug()` output below
+      return new NestedPropertyReference(this, property);
+    } else {
+      return PropertyReference.create(this, property);
+    }
+  }
+}
+
+if (DEBUG) {
+  OutletModelReference.prototype['debug'] = function debug(): string {
+    return '@model';
+  }
+}
+
 class OutletComponentReference
   implements VersionedPathReference<CurriedComponentDefinition | null> {
   public tag: Tag;
-  private definition: CurriedComponentDefinition | null;
-  private lastState: OutletDefinitionState | null;
+  private args: Option<CapturedArguments> = null;
+  private definition: Option<CurriedComponentDefinition> = null;
+  private lastState: Option<OutletDefinitionState> = null;
 
   constructor(private outletRef: VersionedPathReference<OutletState | undefined>) {
-    this.definition = null;
-    this.lastState = null;
     // The router always dirties the root state.
-    this.tag = outletRef.tag;
+    let tag = this.tag = outletRef.tag;
+
+    if (EMBER_ROUTING_MODEL_ARG) {
+      let modelRef = new OutletModelReference(outletRef);
+      let map = dict<VersionedPathReference>();
+      map.model = modelRef;
+
+      // TODO: the functionailty to create a proper CapturedArgument should be
+      // exported by glimmer, or that it should provide an overload for `curry`
+      // that takes `PreparedArguments`
+      this.args = {
+        tag,
+        positional: EMPTY_ARGS.positional,
+        named: {
+          tag,
+          map,
+          names: ['model'],
+          references: [modelRef],
+          length: 1,
+          has(key: string): boolean {
+            return key === 'model';
+          },
+          get<T extends VersionedPathReference>(key: string): T {
+            assert('key !== model', key === 'model');
```

Comment:
Am I correct in assuming that this asserstion is what would be thrown when you do `@foo` in a route's template? If so, it should definitely be made a bit more user friendly...

Reply:
Fixed, it is actually supposed to return `UNDEFINED_REFERENCE`